---
title: "R_Review"
author: "Elizabeth Glynne"
date: "10/21/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
install.packages("tidyverse") #installs the package collection "tidyverse"
library(tidyverse) #loads all packages in "tidyverse"
```

# Downloads genotype and SNP data and assigns them to the appropriate variables
```{r}
fang_et_al_genotypes <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
snp_positions <- read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/snp_position.txt")
```

# Inspecting the structure and dimensions
```{r}
dim(fang_et_al_genotypes) #number of rows x number of columns
dim(snp_positions)
nrow(fang_et_al_genotypes) #number of rows
nrow(snp_positions)
ncol(fang_et_al_genotypes) #number of columns
ncol(snp_positions)
str(fang_et_al_genotypes) #type of object, number of observations and variables, name of variables, variable type, data preview
str(snp_positions)
```

# Filtering maize and teosinte groups from `fang_et_al_genotypes` into separate data frames
```{r}
maize_genotypes <- filter(fang_et_al_genotypes, `Group` == "ZMMIL" | `Group` == "ZMMLR" | `Group` == "ZMMMR") #isolates maize groups from `fang_et_al_genotypes`
teosinte_genotypes <- filter(fang_et_al_genotypes, `Group` == "ZMPBA" | `Group` == "ZMPIL" | `Group` == "ZMPJA")
```

# Transposes maize and teosinte genotype data into new dataframes
```{r}
transposed_maize_genotypes <- as.data.frame(t(as.matrix(maize_genotypes)[,-c(1,2,3)]), stringsAsFactors = FALSE) #transposes all but the first 3 columns
colnames(transposed_maize_genotypes) <- maize_genotypes$Sample_ID #restores column names as Sample_ID from maize genotype data
transposed_maize_genotypes$SNP_ID <- rownames(transposed_maize_genotypes) #adds a variable column with SNP_IDs
rownames(transposed_maize_genotypes) <- NULL #gets rid of SNP_ID rownames
transposed_teosinte_genotypes <- as.data.frame(t(teosinte_genotypes[,-c(1,2,3)]), stringsAsFactors = FALSE) #transposes all but the first 3 columns
colnames(transposed_teosinte_genotypes) <- teosinte_genotypes$Sample_ID #restores column names as Sample_ID from teosinte genotype data
transposed_teosinte_genotypes$SNP_ID <- rownames(transposed_teosinte_genotypes)
rownames(transposed_teosinte_genotypes) <- NULL
```

# Choosing desired columns from `snp_positions` and writing them to a new dataframe
```{r}
reduced_snp_positions <- snp_positions[ ,c("SNP_ID","Chromosome","Position")]
```

# Merge maize and teosinte genotype dataframes with `reduced_snp_positions` dataframe
```{r}
merged_snp_geno_maize <- merge(reduced_snp_positions, transposed_maize_genotypes) #merges `reduced_snp_positions` and `transposed_maize_genotypes` by their common column `$SNP_ID`
merged_snp_geno_teosinte <- merge(reduced_snp_positions, transposed_teosinte_genotypes) #merges `reduced_snp_positions` and `transposed_teosinte_genotypes` by their common column `$SNP_ID`
```

# Replacing "?/?" with "?" or "-"
```{r}
merged_snp_geno_maize_QM <- data.frame(apply(merged_snp_geno_maize, c(1, 2), function(x) {x <- gsub("?/?", "?", x, fixed = TRUE)}), stringsAsFactors = FALSE) #creates a dataframe after applying the "gsub" function to every value in the dataframe `merged_snp_geno_maize` to substitute every instance of "?/?" with "?"
merged_snp_geno_maize_D <- data.frame(apply(merged_snp_geno_maize, c(1, 2), function(x) {x <- gsub("?/?", "-", x, fixed = TRUE)}), stringsAsFactors = FALSE) #creates a dataframe after applying the "gsub" function to every value in the dataframe `merged_snp_geno_maize` to substitute every instance of "?/?" with "-"
merged_snp_geno_teosinte_QM <- data.frame(apply(merged_snp_geno_teosinte, c(1, 2), function(x) {x <- gsub("?/?", "?", x, fixed = TRUE)}), stringsAsFactors = FALSE) #creates a dataframe after applying the "gsub" function to every value in the dataframe `merged_snp_geno_teosinte` to substitute every instance of "?/?" with "?"
merged_snp_geno_teosinte_D <- data.frame(apply(merged_snp_geno_teosinte, c(1, 2), function(x) {x <- gsub("?/?", "-", x, fixed = TRUE)}), stringsAsFactors = FALSE) #creates a dataframe after applying the "gsub" function to every value in the dataframe `merged_snp_geno_teosinte` to substitute every instance of "?/?" with "-"
```

# Parsing dataframes into files based on chromosome number and sorting those files based on SNP position
```{r}
list_merged_files <- list(a=merged_snp_geno_maize_D, b=merged_snp_geno_maize_QM, c=merged_snp_geno_teosinte_D, d=merged_snp_geno_teosinte_QM) #creates a list of maize and teosinte dataframes with either "?" or "-" representing missing data. I made this but couldn't figure out how to use it to loop through the dataframes.

##This function iterates through the 10 chromosomes, filters the initial dataset by each chromosome, sorts them according to the assignment specs, and writes the new dataframe to a new file.
filter_by_chromosome_asc <- function(df)
  {
  for (i in 1:10)
    {
    file_name <- paste("sorted_",deparse(substitute(df)),"_chr_", as.character(i),".txt", sep = "")
    temp_df <- filter(df, df$Chromosome == i)
    sorted_temp_df <- arrange(temp_df, as.numeric(Position))
    write_tsv(sorted_temp_df, file_name)
    }
}
filter_by_chromosome_dsc <- function(df)
  {
  for (i in 1:10)
    {
    file_name <- paste("sorted_",deparse(substitute(df)),"_chr_", as.character(i),".txt", sep = "")
    temp_df <- filter(df, df$Chromosome == i)
    sorted_temp_df <- arrange(temp_df, desc(as.numeric(Position)))
    write_tsv(sorted_temp_df, file_name)
    }
  }

filter_by_chromosome_dsc(merged_snp_geno_maize_D)
filter_by_chromosome_asc(merged_snp_geno_maize_QM)
filter_by_chromosome_asc(merged_snp_geno_teosinte_D)
filter_by_chromosome_dsc(merged_snp_geno_teosinte_QM)
```

# Plotting SNPs per chromosome
```{r}
ggplot(data = merged_snp_geno_maize) + geom_bar(mapping = aes(x = Chromosome))
ggsave("maize_SNPsperChromosome.pdf")

ggplot(data = merged_snp_geno_teosinte) + geom_bar(mapping = aes(x = Chromosome))
ggsave("teosinte_SNPsperChromosome.pdf")

ggplot(data = merged_snp_geno_maize) + geom_point(mapping = aes(x = Position, y = Chromosome), size = 1)
ggsave("maize_SNPdistChromosomes.pdf")

ggplot(data = merged_snp_geno_teosinte) + geom_point(mapping = aes(x = Position, y = Chromosome), size = 1)
ggsave("teosinte_SNPdistChromosomes.pdf")
```

# Indicate whether a site is heterozygous or not.
```{r}
install.packages("reshape2")
library(reshape2)

melted_maize <- melt(merged_snp_geno_maize, id.vars = c("SNP_ID", "Chromosome", "Position"), variable.name = "Sample", value.name = "SNP")
melted_teosinte <- melt(merged_snp_geno_teosinte, id.vars = c("SNP_ID", "Chromosome", "Position"), variable.name = "Sample", value.name = "SNP")

melted_maize_annotated <- mutate(melted_maize, AllelType = case_when((melted_maize$SNP=="A/A" | melted_maize$SNP=="G/G" | melted_maize$SNP=="C/C" | melted_maize$SNP=="T/T") ~ "homozygous", melted_maize$SNP=="?/?" ~ "Unknown", TRUE ~ "heterozygous"))
melted_teosinte_annotated <- mutate(melted_teosinte, AllelType = case_when((melted_teosinte$SNP=="A/A" | melted_teosinte$SNP=="G/G" | melted_teosinte$SNP=="C/C" | melted_teosinte$SNP=="T/T") ~ "homozygous", melted_teosinte$SNP=="?/?" ~ "Unknown", TRUE ~ "heterozygous"))

ggplot(data = melted_maize_annotated) + geom_bar(mapping = aes(x = AllelType))
ggsave("maize_heterozygosity.pdf")
ggplot(data = melted_teosinte_annotated) + geom_bar(mapping = aes(x = AllelType))
ggsave("teosinte_heterozygosity.pdf")
```

#My own thang
```{r}
ggplot(data = melted_maize_annotated) + geom_point(mapping = aes(x = Position, y = Chromosome, color = SNP), size = 1)
ggsave("maize_SNPdistChromosomesSNPcolor.pdf")
ggplot(data = melted_teosinte_annotated) + geom_point(mapping = aes(x = Position, y = Chromosome, color = SNP), size = 1)
ggsave("teosinte_SNPdistChromosomesSNPcolor.pdf")
```

# Comments:
- I think this was well done. It ran through for me with no issues. 
- Could you have removed the Unknown and Multiple groups from your plots?
- Why did you choose to color based on SNP in plotting Position vs Chromosome? I think the addition of color was nice, but it didn't provide additional information to the diagram for me. It would be cool to see them colored with matching bases as one color and mis-matched bases as another. 